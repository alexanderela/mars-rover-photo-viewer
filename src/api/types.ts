import { Obj, Rover } from "../types/common";

export interface FetchRoverProps {
  rover: Rover;
  options?: Obj;
  page?: string;
}

export interface RoverPhotoRaw {
  id: number;
  sol: number;
  camera: RoverPhotoCameraRaw;
  img_src: string;
  earth_date: string;
  rover: RoverPhotoRoverRaw
}

export interface RoverPhotoRoverRaw {
  id: number;
  name: string;
  landing_date: string;
  launch_date: string;
  status: string;
  max_sol: number;
  max_date: string;
  total_photos: number;
  cameras: RoverPhotoRoverCameraRaw[];
}

export interface RoverPhotoRoverCameraRaw {
  name: string;
  full_name: string;
}

export interface RoverPhotoCameraRaw {
  id: number;
  name: string;
  rover_id: number;
  full_name: string;
}