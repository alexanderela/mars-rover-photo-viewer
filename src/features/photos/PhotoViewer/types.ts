import { Rover, RoverPhotoStateObj } from "../../../types/common";

export interface RoverPhotoState {
  photos: RoverPhotoStateObj;
  selectedRover: Rover;
  page: number;
}