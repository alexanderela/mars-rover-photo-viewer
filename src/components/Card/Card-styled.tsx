import { IconButton } from "@mui/material";
import { styled } from "@mui/material/styles";

export const StyledFavoriteButton = styled(IconButton)(({ theme }) => ({
  outline: "none",
  position: "absolute",
  right: theme.spacing(0.25),
  bottom: theme.spacing(0.25),
  color: "oldlace",
  zIndex: 10
}));