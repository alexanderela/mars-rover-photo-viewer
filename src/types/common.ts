export interface Obj {
  [key: string]: string;
}

export enum Rover {
  CURIOSITY = "curiosity",
  OPPORTUNITY = "opportunity",
  SPIRIT = "spirit",
}

export interface RoverPhoto {
  id: number;
  name: string;
  roverName: string;
  imgSrc: string;
  earthDate: string;
  sol: number;
  cameraName: string;
  cameraFullName: string;
}

// export interface RoverPhotoStateObj {
//   [id: number]: RoverPhoto;
// }
export interface RoverPhotoStateObj {
  [id: number]: RoverPhoto;
}
