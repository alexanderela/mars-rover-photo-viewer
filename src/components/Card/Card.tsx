import { Box, useTheme } from "@mui/material";
import { RoverPhoto } from "../../types/common";
import { memo, useCallback } from "react";
import { useNavigate, useSearchParams } from "react-router-dom";

interface CardProps {
  photo: RoverPhoto;
}

export const Card = memo(function Card({ photo }: CardProps) {
  const theme = useTheme();
  const navigate = useNavigate();
  const [searchParams] = useSearchParams();
  const page = searchParams.get("page");

  const handleCardClick = useCallback(() => {
    navigate(`id/${photo.id}?page=${parseInt(page as string)}`);
  }, [navigate, page, photo.id]);

  return (
    <Box
      onClick={handleCardClick}
      sx={{
        borderRadius: theme.spacing(1),
        boxShadow: `0px 0px 0px ${theme.spacing(0.25)} white inset`,
        height: theme.spacing(18.8525),
        width: theme.spacing(25),
        margin: `${theme.spacing(1)} ${theme.spacing(2)}`,
        backgroundImage: `url(${photo.imgSrc})`,
        backgroundSize: "cover",
        cursor: "pointer"
      }}
      data-testid="Card"
    />
  );
});