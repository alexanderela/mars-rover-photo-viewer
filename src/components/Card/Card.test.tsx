import { describe, expect, it } from "vitest";
import { Card } from "./Card";
import { render, screen, fireEvent } from "../../tests/test-utils";
import { RoverPhoto } from "../../types/common";
import { createMemoryRouter, MemoryRouter, RouterProvider } from "react-router-dom";
import { act } from "react";

const mockPhoto: RoverPhoto = {
  id: 1228204,
  name: "Curiosity",
  roverName: "curiosity",
  imgSrc: "https://mars.nasa.gov/msl-raw-images/msss/04102/mhli/4102MH0001530001404334U01_DXXX.jpg",
  earthDate: "2024-02-19",
  sol: 4102,
  cameraName: "MAHLI",
  cameraFullName: "Mars Hand Lens Imager"
}

describe("<Card />", () => {
  it("renders Card component", () => {
    render(
      <MemoryRouter>
        <Card photo={mockPhoto} />
      </MemoryRouter>
    );
    const element = screen.getByTestId(/Card/i);
    expect(element).toBeInTheDocument();
    expect(element).toMatchSnapshot();
  });

  it("updates route to /id/:id upon clicking photo", () => {
    const routes = [
      {
        path: "/rovers/curiosity",
        element: <Card photo={mockPhoto} />
      }
    ]
    const router = createMemoryRouter(routes, {
      initialEntries: ["/", "/rovers/curiosity"],
      initialIndex: 1
    });
    const result = render(<RouterProvider router={router} />);
    const element = result.getByTestId(/Card/i);
    expect(router.state.location.pathname).toEqual("/rovers/curiosity")
    act(() => {
      fireEvent.click(element);
    })
    expect(router.state.location.pathname).toEqual("/rovers/curiosity/id/1228204");
  });
});