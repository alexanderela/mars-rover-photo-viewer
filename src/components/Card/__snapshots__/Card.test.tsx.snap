// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Card /> > renders Card component 1`] = `
<div
  class="MuiBox-root css-1x1ddkk"
  data-testid="Card"
/>
`;
